src/fundus/scraping/crawler.py

```diff
@@ -106,10 +113,213 @@ def pool_queue_iter(handle: MapResult[Any], queue: Queue[_T]) -> Iterator[_T]:
+    def __init__(
+        self,
+        *publishers: Publisher,
+        restrict_sources_to: Optional[List[Type[URLSource]]] = None,
+        delay: Optional[Union[float, Delay]] = 1.0,
+        threading: bool = True,
+    ):
+        """Fundus base class for crawling articles from the web.
+
+        Examples:
+            >>> from fundus import PublisherCollection, Crawler
+            >>> crawler = Crawler(*PublisherCollection)
+            >>> # Crawler(PublisherCollection.us) to crawl only english news
+            >>> for article in crawler.crawl():
+            >>>     print(article)
+
+        Args:
+            *publishers (Union[PublisherEnum, Type[PublisherEnum]]): The publishers to crawl.
+            restrict_sources_to (Optional[List[Type[URLSource]]]): Lets you restrict
+                sources defined in the publisher specs. If set, only articles from given source types
+                will be yielded.
+            delay (Optional[Union[float, Delay]]): Set a delay time in seconds to be used between article
+                downloads. You can set a delay directly using float or any callable satisfying the Delay
+                protocol. If set to None, no delay will be used between batches. See Delay for more
+                information. Defaults to None.
+        """
+
+        if not publishers:
+            raise ValueError("param <publishers> of <Crawler.__init__> has to be non empty")
+
+        super().__init__(*publishers)
+
+        self.restrict_sources_to = restrict_sources_to
+        self.delay = delay
+        self.threading = threading
+
+    def _fetch_articles(
+        self,
+        publisher: PublisherEnum,
+        error_handling: Literal["suppress", "catch", "raise"],
+        extraction_filter: Optional[ExtractionFilter] = None,
+        url_filter: Optional[URLFilter] = None,
+    ) -> Iterator[Article]:
+        def build_delay() -> Optional[Delay]:
+            if isinstance(self.delay, float):
+
+                def constant_delay() -> float:
+                    return self.delay  # type: ignore[return-value]
+
+                return constant_delay
+
+            elif isinstance(self.delay, Delay):
+                return self.delay
+
+            else:
+                raise TypeError("param <delay> of <Crawler.__init__>")
+
+        scraper = WebScraper(publisher, self.restrict_sources_to, build_delay())
+        yield from scraper.scrape(error_handling, extraction_filter, url_filter)
+
+    @staticmethod
+    def _single_crawl(
+        publishers: Tuple[PublisherEnum, ...], article_task: Callable[[PublisherEnum], Iterator[Article]]
+    ) -> Iterator[Article]:
+        article_iterators = [article_task(publisher) for publisher in publishers]
+        while article_iterators:
+            for iterator in article_iterators:
+                try:
+                    yield next(iterator)
+                except StopIteration:
+                    article_iterators.remove(iterator)
```

Comment:
> Isn't the same issue still there, since tmp is just a direct reference to article_iterators?

Yes, that's true. I overlooked that i would have needed any form of `copy`, which makes this somehow inefficient.

> I think using [interleave_longest](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.interleave_longest) or [roundrobin](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.roundrobin) from more itertools would help here.

Yes. I still had in my mind that both share a significant flaw that requires one element from each iterator to be retrieved before yielding the first one, making it unusable for our case, but that's only the case for `interleave_longest`. I'm gonna replace the existing code with `roundrobin.`

Reply:
> I still had in my mind that both share a significant flaw that requires one element from each iterator to be retrieved before yielding the first one

Interesting, I did not know that. Sounds like a weird implementation detail. Great, that roundrobin works.